### PR TITLE
Add smoke tests for spack install

### PR DESCRIPTION
**Pull request by G-Ragghianti**
We need to include the required "smoke test" files so that spack can use these to validate the installation of papi.  I put these into `src/smoke_tests`.  Is there a better location?

Once the location and require test environment are set, we can implement the smoke tests within the papi spack package.

### CODE REVIEW

**Comment by gcongiu** [comment]
Probably stupid question. Why don't we simply run the PAPI utilities instead? E.g., `papi_command_line`, `papi_component_avail`, etc?

**Comment by G-Ragghianti** [comment]
The smoke tests satisfy a different need.  They are intended to replicate
what a user would do after the papi package has been installed.  In this
case, papi_component_avail and similar could be used, but the main interest
is being able to link other code against the installed papi library and
make sure that the code can minimarlly use the functionality of the library.

**Comment by gcongiu** [comment]
Ok, since the utilities are programs statically linked to `libpapi.a` this makes sense. Would it also make sense to run, along with smoke tests, some of the utilities too? The case for `papi_component_avail`, for example, is that it will force the initialisation of all components. Thus, it is a further guarantee that all components will at least initialise correctly. The smoke tests don't do that as far as I can tell...

> The smoke tests satisfy a different need. They are intended to replicate what a user would do after the papi package has been installed. In this case, papi_component_avail and similar could be used, but the main interest is being able to link other code against the installed papi library and make sure that the code can minimarlly use the functionality of the library.
> […](#)

**Comment by G-Ragghianti** [comment]
Yes, we can also develop a script that runs tests on the installed
utilities.

File: .github/workflows/spack.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export HOME=`pwd`
+git clone https://github.com/spack/spack spack || true
+source spack/share/spack/setup-env.sh
+
+SPEC="papi@master +lmsensors +powercap +sde +infiniband +rapl +cuda %$COMPILER"
+echo SPEC=$SPEC
+
+module load $COMPILER
+spack compiler find
+#spack uninstall -a -y papi || true
```
Comment (gcongiu) [inline .github/workflows/spack.sh:22]: Can you remove this comment?

File: src/smoke_tests/Makefile [hunk 2]
```diff
@@ -0,0 +1,22 @@
+
+EXECUTABLES = simple threads
+
+CC ?= gcc
+
+PAPIROOT ?= /path/to/papi/install/prefix
+CFLAGS ?= -O0 -pthread -I$(PAPIROOT)/include
```
Comment (gcongiu) [inline src/smoke_tests/Makefile:7]: Applications will likely use -O2

File: src/smoke_tests/threads.c [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        printf( "%12lld \t\t --> %s  (thread %d) \n", values[i],
+                EventName[i], thread );
+    }
+
+    retval = PAPI_unregister_thread(  );
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_unregister_thread: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }
+
```
Comment (gcongiu) [inline src/smoke_tests/threads.c:72]: Should destroy the eventset too.

File: src/smoke_tests/simple.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_stop: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }
+                
+    for( i = 0; i < NUM_EVENTS; i++ ) {           
+        printf( "%12lld \t\t --> %s  \n", values[i],
+                EventName[i] );
+    }
+
```
Comment (gcongiu) [inline src/smoke_tests/simple.c:65]: Should destroy eventset too.

**Review by gcongiu — CHANGES_REQUESTED** [review]
Looks good to me. Just a couple minor comments. For the utilities we should check with @jagode she is fine adding some (e.g., `papi_component_avail`). I would have them but maybe is not necessary.

When everything is finalised could you also cleanup the commits, i.e. squash the ones that need squashing, and adding editing comments using the format in the wiki (Authors checklist)?

File: src/smoke_tests/simple.c (hunk 4 above)
Comment (jagode) [reply src/smoke_tests/simple.c:65]: AFAIK, PAPI_shutdown(), which I included, cleans up and frees event sets.

File: src/smoke_tests/threads.c (hunk 3 above)
Comment (jagode) [reply src/smoke_tests/threads.c:72]: AFAIK, PAPI_shutdown(), which I included, cleans up and frees event sets.

**Review by jagode — APPROVED** [review]

**Comment by jagode** [comment]
> Looks good to me. Just a couple minor comments. For the utilities we should check with @jagode she is fine adding some (e.g., `papi_component_avail`). I would have them but maybe is not necessary.

Not needed for the spack smoke tests.

File: .github/workflows/spack.sh (hunk 1 above)
Comment (G-Ragghianti) [reply .github/workflows/spack.sh:22]: yep

File: src/smoke_tests/Makefile (hunk 2 above)
Comment (G-Ragghianti) [reply src/smoke_tests/Makefile:7]: This was provided by Heike.  I don't have a preference.

**Comment by G-Ragghianti** [comment]
> Looks good to me. Just a couple minor comments. For the utilities we should check with @jagode she is fine adding some (e.g., `papi_component_avail`). I would have them but maybe is not necessary.
> 
> When everything is finalised could you also cleanup the commits, i.e. squash the ones that need squashing, and adding editing comments using the format in the wiki (Authors checklist)?

I rebased according to the wiki instructions.  Squash which commits?  Add the checklist to the original PR content or to a comment?  Do I need to go back and change the commit headers and bodies according to the checklist?

**Comment by gcongiu** [comment]
I think we can squash commits into only two commits: one that adds the C smoke tests and Makefile (commits 6 and 7) and one that concerns the CI integration of those tests (all the remaining).

**Comment by G-Ragghianti** [comment]
OK I'll see about that.  Just a detail: the smoke tests are not part of the CI.  They are to be used by the spack package for papi when tested after install via spack (mainly required by ECP projects).

**Review by gcongiu — APPROVED** [review]

**Pull request merged**